drivers/clock_control/clock_control_renesas_cpg_mssr.c

```diff
@@ -58,3 +58,319 @@ int rcar_cpg_mstp_clock_endisable(uint32_t base_address, uint32_t module, bool e
+		children_list->in_freq = parent->out_freq;
+
+		if (rcar_cpg_update_out_freq(dev, children_list) < 0) {
+			__ASSERT_NO_MSG(0);
+			return;
+		}
+
+		/* child can have childrens */
+		rcar_cpg_change_children_in_out_freq(dev, children_list);
+		children_list = children_list->next_sibling;
+	}
+}
+
+int rcar_cpg_get_rate(const struct device *dev, clock_control_subsys_t sys, uint32_t *rate)
+{
+	int ret;
+	struct rcar_cpg_mssr_data *data;
+	struct rcar_cpg_clk *clk = (struct rcar_cpg_clk *)sys;
+	k_spinlock_key_t key;
+
+	struct cpg_clk_info_table *clk_info;
+
+	if (!dev || !sys || !rate) {
+		LOG_ERR("%s: received null ptr input arg(s) dev %p sys %p rate %p",
+			__func__, dev, sys, rate);
+		return -EINVAL;
+	}
+
+	clk_info = rcar_cpg_find_clk_info_by_module_id(dev, clk->domain, clk->module);
+	if (clk_info == NULL) {
+		return -EINVAL;
+	}
+
+	data = dev->data;
+
+	key = k_spin_lock(&data->lock);
+	ret = rcar_cpg_get_out_freq(dev, clk_info);
+	k_spin_unlock(&data->lock, key);
+
+	if (ret < 0) {
+		return ret;
+	}
+
+	*rate = ret;
+	return 0;
+}
+
+int rcar_cpg_set_rate(const struct device *dev, clock_control_subsys_t sys,
+		      clock_control_subsys_rate_t rate)
+{
+	int ret = -ENOTSUP;
+	k_spinlock_key_t key;
+	struct cpg_clk_info_table *clk_info;
+	struct rcar_cpg_clk *clk = (struct rcar_cpg_clk *)sys;
+	struct rcar_cpg_mssr_data *data;
+	int in_freq;
+	uint32_t divider;
+	uint32_t div_mask;
+	uint32_t module;
+	uintptr_t u_rate = (uintptr_t)rate;
+
+	if (!dev || !sys || !rate) {
+		LOG_ERR("%s: received null ptr input arg(s) dev %p sys %p rate %p",
+			__func__, dev, sys, rate);
+		return -EINVAL;
+	}
+
+	clk_info = rcar_cpg_find_clk_info_by_module_id(dev, clk->domain, clk->module);
+	if (clk_info == NULL) {
+		return -EINVAL;
+	}
+
+	if (clk_info->domain == CPG_MOD) {
+		__ASSERT_NO_MSG(clk_info->parent);
+		clk_info = clk_info->parent;
+	}
+
+	module = clk_info->module;
+	data = dev->data;
+
+	key = k_spin_lock(&data->lock);
+	in_freq = rcar_cpg_get_in_update_out_freq(dev, clk_info);
+	if (in_freq < 0) {
+		ret =  in_freq;
+		goto unlock;
+	}
+
+	if (u_rate == 0) {
+		ret = -EINVAL;
+		goto unlock;
+	}
+
+	divider = in_freq / u_rate;
+	if (divider * u_rate != in_freq) {
+		ret = -EINVAL;
+		goto unlock;
+	}
+
+	if (!data->set_rate_helper) {
+		ret = -ENOTSUP;
+		goto unlock;
+	}
+
+	ret = data->set_rate_helper(module, &divider, &div_mask);
+
+	if (ret < 0) {
+		uint32_t reg = sys_read32(clk_info->offset + data->base_addr);
+
+		reg &= ~div_mask;
+		rcar_cpg_write(data->base_addr, clk_info->offset, reg | divider);
+		clk_info->out_freq = u_rate;
+
+		rcar_cpg_change_children_in_out_freq(dev, clk_info);
+	}
+
+unlock:
+	k_spin_unlock(&data->lock, key);
+	return ret;
+}
+
+void rcar_cpg_build_clock_relationship(const struct device *dev)
+{
+	uint32_t domain;
+	k_spinlock_key_t key;
+	struct rcar_cpg_mssr_data *data = dev->data;
+
+	if (!data) {
+		return;
+	}
+
+	key = k_spin_lock(&data->lock);
+	for (domain = 0; domain < 2; domain++) {
```

Comment:
Why exactly `2`?

Reply:
we are going through tables of domains here and it is possible to have only two tables: CORE and MOD